File: dependency_injection/service_collection.py
from enum import Enum, auto


class ServiceLifetime(Enum):
    TRANSIENT = auto()
    SINGLETON = auto()
    INSTANCE = auto()


class ContainerService:
    def __init__(self, service_implementation, service_lifetime, args=None):
        self.service_implementation = service_implementation
        self.service_lifetime = service_lifetime
        self.args = args
        self.instance = None


class ServiceCollection:
    def __init__(self):
        self._service_collection = dict()

    def add(self, service_to_add, service_implementation, service_lifetime, args=None):
        if service_implementation is None:
            service_implementation = service_to_add

        self._service_collection[service_to_add.__name__] = ContainerService(
            service_implementation, service_lifetime, args
        )

    def add_transient(self, service_to_add, service_implementation=None, args=None):
        self.add(service_to_add, service_implementation, ServiceLifetime.TRANSIENT, args)

    def resolve(self, service_to_resolve):
        container_service = self._service_collection[service_to_resolve.__name__]
        if container_service.args:
            return container_service.service_implementation(*container_service.args)
        annotations = container_service.__init__.__annotations__
        services_to_use = list()
        for annotation in annotations.values():
            services_to_use.append(self.resolve(annotation))
        return container_service.service_implementation(*services_to_use)

File: src/ServiceCollection.cpp
#include <pybind11/pybind11.h>
#include "ServiceCollection.h"
#include "ContainerService.h"
#include "ServiceLifetime.h"
#include "Exceptions.h"

using namespace pybind11::literals;

py::object _ServiceCollection::Resolve(py::object service)
{
    ssize_t serviceId = py::hash(service);
    auto it = _serviceCollection.find(serviceId);
    if (it == _serviceCollection.end())
    {
        throw ServiceResolutionError("Requested service not found in collection.", service);
    }
    ContainerService &serviceToResolve = _serviceCollection[serviceId];
    if (serviceToResolve.Lifetime == ServiceLifetime::INSTANCE)
        return ResolveInstance(serviceToResolve);
    if (serviceToResolve.Lifetime == ServiceLifetime::SINGLETON)
        return ResolveSingleton(serviceToResolve);
    return ResolveTransient(serviceToResolve);
}

void _ServiceCollection::Add(py::object service, py::object serviceImplementation, ServiceLifetime lifetime, std::map<std::string, py::object> annotations)
{
    ssize_t serviceId = py::hash(service);
    _serviceCollection[serviceId] = ContainerService(serviceImplementation, lifetime, annotations);
}

void _ServiceCollection::Add(py::object service, py::object serviceImplementation, ServiceLifetime lifetime, std::map<std::string, py::object> annotations, py::list args)
{
    ssize_t serviceId = py::hash(service);
    _serviceCollection[serviceId] = ContainerService(serviceImplementation, lifetime, annotations, args);
}

py::object _ServiceCollection::ResolveAnnotations(ContainerService &containerService)
{
    if (!containerService.Args.empty())
    {
        return ResolveArgs(containerService);
    }
    auto implementation = containerService.Implementation;
    auto annotations = containerService.Annotations;
    std::vector<py::object> services_to_use;
    for (const auto &[name, annotation] : annotations)
    {

        if (name != "return")
        {
            services_to_use.push_back(Resolve(annotation));
        }
    }
    return implementation(*py::cast(services_to_use));
}

py::object _ServiceCollection::ResolveArgs(ContainerService &containerService)
{
    return containerService.Implementation(*containerService.Args);
}

py::object _ServiceCollection::ResolveSingleton(ContainerService &containerService)
{
    if (containerService.SingletonInstance.is_none())
    {
        containerService.SingletonInstance = ResolveAnnotations(containerService);
    }
    return containerService.SingletonInstance;
}

py::object _ServiceCollection::ResolveTransient(ContainerService &containerService)
{
    return ResolveAnnotations(containerService);
}

py::object _ServiceCollection::ResolveInstance(ContainerService &containerService)
{
    return containerService.Implementation;
}
